gnomad_qc/v4/sample_qc/assign_ancestry.py

```diff
@@ -0,0 +1,458 @@
+            withheld_sample=pop_pca_scores_ht[pop_ht.key].withheld_sample
+        )
+
+    return pop_ht, pops_rf_model
+
+
+def write_pca_results(
+    pop_pca_eigenvalues: List[float],
+    pop_pca_scores_ht: hl.Table,
+    pop_pca_loadings_ht: hl.Table,
+    overwrite: hl.bool,
+    removed_unreleasables: hl.bool,
+    test: hl.bool = False,
+):
+    """
+    Write out the three objects returned by run_pca().
+
+    :param pop_pca_eigenvalues: List of eigenvalues returned by run_pca.
+    :param pop_pca_scores_ht: Table of scores returned by run_pca.
+    :param pop_pca_loadings_ht: Table of loadings returned by run_pca.
+    :param overwrite: Whether to overwrite an existing file.
+    :param removed_unreleasables: Whether run_pca removed unreleasable samples.
+    :param test: Whether the test qc mt was used in pca.
+    :return: None
+    """
+    pop_pca_eigenvalues_ht = hl.Table.parallelize(
+        hl.literal(
+            [{"PC": i + 1, "eigenvalue": x} for i, x in enumerate(pop_pca_eigenvalues)],
+            "array<struct{PC: int, eigenvalue: float}>",
+        )
+    )
+    pop_pca_eigenvalues_ht.write(
+        ancestry_pca_eigenvalues(removed_unreleasables, test).path,
+        overwrite=overwrite,
+    )
+    pop_pca_scores_ht.write(
+        ancestry_pca_scores(removed_unreleasables, test).path,
+        overwrite=overwrite,
+    )
+    pop_pca_loadings_ht.write(
+        ancestry_pca_loadings(removed_unreleasables, test).path,
+        overwrite=overwrite,
+    )
+
+
+def main(args):
+    hl.init(
+        log="/assign_ancestry.log",
+        default_reference="GRCh38",
+        tmp_dir="gs://gnomad-tmp-4day",
+    )
+
+    remove_unreleasables = args.remove_unreleasable_samples
+    overwrite = args.overwrite
+    test = args.test
+
+    if args.run_pca:
+        pop_eigenvalues, pop_scores_ht, pop_loadings_ht = run_pca(
+            remove_unreleasables,
+            args.n_pcs,
+            pca_related_samples_to_drop.ht(),
+            test,
+        )
+
+        write_pca_results(
+            pop_eigenvalues,
+            pop_scores_ht,
+            pop_loadings_ht,
+            overwrite,
+            remove_unreleasables,
+            test,
+        )
+
+    if args.assign_pops:
+        pop_pcs = args.pop_pcs
+        pop_ht, pops_rf_model = assign_pops(
+            args.min_pop_prob,
+            remove_unreleasables,
+            max_number_mislabeled_training_samples=args.max_number_mislabeled_training_samples,
+            max_proportion_mislabeled_training_samples=args.max_proportion_mislabeled_training_samples,
+            pcs=pop_pcs,
+            withhold_prop=args.withhold_prop,
+            test=test,
+            overwrite=overwrite,
+        )
+        pop_ht = pop_ht.checkpoint(
+            get_pop_ht(test=test).path,
+            overwrite=overwrite,
+            _read_if_exists=not overwrite,
+        )
+        pop_ht.transmute(
+            **{f"PC{i}": pop_ht.pca_scores[i - 1] for i in pop_pcs}
```

Comment:
but the scores actually may not be grabbed properly...

Reply:
The scores are a list so PC1 is in 0, PC8 is in 7, so they are retrieved properly

Reply:
just want to double check this, because there's a step within assign_population_pcs that removes the unnecessary PCs in the array (https://github.com/broadinstitute/gnomad_methods/blob/2edccf5dbab8b73a2e6f249b470e1ef80a765b24/gnomad/sample_qc/ancestry.py#L170), so the indexes might not match i-1 anymore

Reply:
So I think the linked line of code is selecting the passed PCs out of the scores HT and uses the same logic as this dictionary value. This function is also assigning the key using the ith entry -- but maybe I am missing something?